README
NAME
    MooseX::Role::AttributeOverride - Allow roles to modify attributes

VERSION
      This document describes v0.0.8 of MooseX::Role::AttributeOverride - released June 22, 2011 as part of MooseX-Role-AttributeOverride.

SYNOPSIS
        {
            package MyApp::Role;
            use Moose::Role;
            use MooseX::Role::AttributeOverride;

            has_plus 'fun' => ( default => 'yep', );

            has_plus 'alive' => (
                default => 'yep',
                override_ignore_missing => 1,
            );
        }
        {
            package MyApp::Trait;
            use Moose::Role;
            use MooseX::Role::AttributeOverride;

            has_plus default => (
                default => sub {
                    my $attr = shift;
                    return sub { $attr->name }
                }
            );
        }
        {
            package MyApp;
            use Moose 1.9900;

            has nolife => (
                is     => 'rw',
                isa    => 'Str',
                traits => ['MyApp::Trait'],
            );

            has 'fun' => (
                is  => 'rw',
                isa => 'Str'
            );

            with qw(MyApp::Role);
        }
        {
            package main;
            use feature 'say';

            my $test = MyApp->new();

            say "I have " . $test->nolife;
            # Says I have nolife
            say "Are you having fun? " . $test->fun;
            # Says Are you having fun? yep

        }

DESCRIPTION
    Moose doesn't allow roles to override attributes using the has '+attr'
    method. There are several good reasons for that. Basically, "that's not
    what a role is for." A role is a set of requirements with defaults. A
    class should always be able to override a role.

    But sometimes you want a role to add features to a class. This is why
    Moose has method modifiers. This extension adds attribute modifiers.

INTERFACE
    has_plus
        This has exactly the same syntax as the Moose has command, except
        you should not use a plus to indicate you are overriding an
        attribute.

    has_plus options

        override_ignore_missing
            Setting this to a true value will allow your role to have
            modifications to attributes that may not exist in the class it
            is applied to. The default is to die in these cases.

            For example:

                package MyApp::Role;
                use Moose::Role;
                use MooseX::Role::AttributeOverride;

                has_plus 'alive' => (
                    default => 'yep',
                    override_ignore_missing => 1,
                );

                package MyApp;
                use Moose;

                with qw(MyApp::Role);
                # I'm not dead yet.

            The above would not die, even though the MyApp package has no
            attribute named 'fun.'

IMPORTANT NOTE
    Always apply a role that uses this module after defining attributes.

META USAGE
    This role can be used in traits. For example, the following works:

        {
            package MyApp::Trait;
            use Moose::Role;
            use MooseX::Role::AttributeOverride;

            has_plus default => (
                default => sub {
                    my $attr = shift;
                    return sub { $attr->name }
                }
            );
        }
        {
            package MyApp;
            use Moose 1.9900;

            has nolife => (
                is     => 'rw',
                isa    => 'Str',
                traits => ['MyApp::Trait'],
            );

            with qw(MyApp::Role);
        }
        {
            package main;
            use feature 'say';

            my $test = MyApp->new();

            say "I have " . $test->nolife;
            # Says I have nolife
        }

DIAGNOSTICS
    Can't find attribute $attr required by $role
        You will see this error if your role has an attribute modification
        for an attribute that is not in the class. You can squash this by
        setting the 'override_ignore_missing' option in your 'has_plus'
        command.

    Attempt to call has_plus on an invalid object
        You really should never see this. Please file a bug report if you
        do. A test case would be nice as well.

    Illegal inherited options
        Moose will throw this error if you try to change an accessor option.
        See the Moose manual for more details.

    Do not use a plus prefix with the has_plus sugar
        There is no need for a plus sign on your attribute:

            # Good
            has 'children', trait => ['good']

            # Bad. Will die.
            has '+children', trait => ['naughty']

DEPENDENCIES
    Moose 1.9900 or newer. Older versions may be supported in a future
    version of this module.

INCOMPATIBILITIES
    I am sure that there are some MooseX modules that will not work with
    this. Please let me know, and I will at least document them.

BUGS AND LIMITATIONS
    This is not the intended use of roles. As a result, take into account
    the following:

    *   Order matters! If two roles modify the same attribute in the same
        way, the second one applied will be the one that is used. This
        behavior, however, relies on Moose keeping track of order, which it
        generally does a good job of, but no guarantees.

    *   Currently, the value of the attribute is clobbered when the role is
        applied. This may change in the future.

    *   This works the same as '+has'. This means that you can't override
        accessor methods. This is a very sensible Moose limitation.

    *   After having an issue with Moose, clone_and_inherit_options, and
        traits that use _process_options, I reimplemented
        clone_and_inherit_optiosn in a way that fixes it. Sort of. A side
        effect of this is that has_plus will not allow you to override the
        lazy option, without a default or builder option.

    *   If you try adding this role before adding the attributes, it won't
        work.

    I am relatively new to Moose. I had an itch, and wrote this Module to
    scratch it. Please let me know how to make this module better.

    For bugs, test cases are great!

SEE ALSO
    Please see those modules/websites for more information related to this
    module.

    *   Moose

    *   Moose::Manual::Attributes

SUPPORT
  Websites
    The following websites have more information about this module, and may
    be of help to you. As always, in addition to those websites please use
    your favorite search engine to discover more resources.

    *   Search CPAN

        The default CPAN search engine, useful to view POD in HTML format.

        <http://search.cpan.org/dist/MooseX-Role-AttributeOverride>

    *   CPAN Testers Dependencies

        The CPAN Testers Dependencies is a website that shows a chart of the
        test results of all dependencies for a distribution.

        <http://deps.cpantesters.org/?module=MooseX::Role::AttributeOverride
        >

    *   CPANTS

        The CPANTS is a website that analyzes the Kwalitee ( code metrics )
        of a distribution.

        <http://cpants.perl.org/dist/overview/MooseX-Role-AttributeOverride>

    *   CPAN Testers

        The CPAN Testers is a network of smokers who run automated tests on
        uploaded CPAN distributions.

        <http://www.cpantesters.org/distro/M/MooseX-Role-AttributeOverride>

    *   CPAN Testers Matrix

        The CPAN Testers Matrix is a website that provides a visual way to
        determine what Perls/platforms PASSed for a distribution.

        <http://matrix.cpantesters.org/?dist=MooseX-Role-AttributeOverride>

  Email
    You can email the author of this module at "EALLENIII at cpan.org"
    asking for help with any problems you have.

  Bugs / Feature Requests
    Please report any bugs or feature requests by email to
    "bug-moosex-role-attributeoverride at rt.cpan.org", or through the web
    interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooseX-Role-AttributeOve
    rride>. You will be automatically notified of any progress on the
    request by the system.

  Source Code
    The code is open to the world, and available for you to hack on. Please
    feel free to browse it and play with it, or whatever. If you want to
    contribute patches, please send me a diff or prod me to pull from your
    repository :)

    <http://github.com/riemann42/MooseX-Role-AttributeOverride>

      git clone git://github.com/riemann42/MooseX-Role-AttributeOverride.git

AUTHOR
    Edward Allen <ealleniii_at_cpan_dot_org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Edward J. Allen III.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.



Changes
==============================
9999-99-99 99:99:99 +0000 HEAD
==============================

  commit 847ef6287080048043909c469e9a18b0d45b13a8
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Wed Jun 22 19:23:30 2011 -0700
  
    Changed prereqs and changed composition_class_roles to be done with
    around method modifier

  commit 2abcba6dbdf0efb737da91b9008e03d858f38357
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Thu Jun 16 16:23:13 2011 -0700
  
    Changed meta.t so it no longer assumes time is always increasing 

================================
2011-06-13 20:32:54 -0700 v0.0.7
================================

  commit 009e83b11309e36e0b53b741319e32c678455508
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Mon Jun 13 20:32:54 2011 -0700
  
    Revised POD. Added notes about placement and meta usage. Also revised
    tests and added meta test.

  commit 4e4ba9c085fdb7f49beacbdc76b2b30a1fd6c1a5
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Fri Jun 10 14:04:35 2011 -0700
  
    Revised tests to include more situations 

================================
2011-06-09 18:10:47 -0700 v0.0.6
================================

  commit 3b153022edd3e0b0706119fae37dfec0e5c92f83
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Thu Jun 9 12:36:18 2011 -0700
  
    playing with dist.ini

  commit 9b847a5dd655e99ae61891070efa5393381edfd8
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Thu Jun 9 11:13:33 2011 -0700
  
    Fixed some silly critic issues on stolen code

  commit 89dee4ede69e17d966ee0755c7ba73bd01bf9c17
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Thu Jun 9 11:02:50 2011 -0700
  
    Changes to devel.ini

  commit 2b7442fb019cca1c5a94f35bca7bf270eebd0cb5
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Thu Jun 9 11:01:32 2011 -0700
  
    Replaced clone_and_inherit with my own hack

  commit d23d0f55f4639116666631e2d788108b5bcbf46d
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Tue Jun 7 22:13:29 2011 -0700
  
    Added trait test. Which fails. 

================================
2011-06-06 21:30:01 -0700 v0.0.5
================================

  commit 7bd514571a5834717a66bdf092c2d20e693e8e96
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Mon Jun 6 21:30:01 2011 -0700
  
    Minor changes to dist.ini

  commit ca218744a83e5351e6ab7d6bcde7cacfad00d52c
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Mon Jun 6 21:17:43 2011 -0700
  
    Configured to use git log for changes as I am lazy 

================================
2011-06-06 21:07:32 -0700 v0.0.4
================================

  commit 4c46432a85c8c641221f7cade20277ccdacd7841
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Mon Jun 6 21:07:32 2011 -0700
  
    New CPAN release of MooseX-Role-AttributeOverride - v0.0.4

  commit c726b7e3d4791118ec52b08591da2e5132b9c0d2
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Mon Jun 6 20:59:32 2011 -0700
  
    Added some author tests and revised POD generation

  commit b46fbd47997a55eaee622cfbcc04ac31eb8fbb97
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Mon Jun 6 20:18:29 2011 -0700
  
    Perl Critic Fixes for silly little things

  commit 803ed79263a3e99d5f1fd7abb354422f6ec9c890
  Author: Edward Allen <ealleniii@cpan.org>
  Date:   Sun Jun 5 23:17:00 2011 -0700
  
    Switched to using Dist::Zilla 

===================================
End of changes in the last 365 days
===================================


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'Module::Build' => '0.3601',
    'Moose::Util' => '0',
    'Test::More' => '0.88',
    'Try::Tiny' => '0'
  },
  'configure_requires' => {
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'Allow roles to modify attributes',
  'dist_author' => [
    'Edward Allen <ealleniii_at_cpan_dot_org>'
  ],
  'dist_name' => 'MooseX-Role-AttributeOverride',
  'dist_version' => '0.0.8',
  'license' => 'perl',
  'module_name' => 'MooseX::Role::AttributeOverride',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'Moose' => '1.9900',
    'perl' => '5.008'
  },
  'script_files' => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Allow roles to modify attributes',
  'AUTHOR' => 'Edward Allen <ealleniii_at_cpan_dot_org>',
  'BUILD_REQUIRES' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'ExtUtils::MakeMaker' => '6.30',
    'Moose::Util' => '0',
    'Test::More' => '0.88',
    'Try::Tiny' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.30'
  },
  'DISTNAME' => 'MooseX-Role-AttributeOverride',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'MooseX::Role::AttributeOverride',
  'PREREQ_PM' => {
    'Moose' => '1.9900'
  },
  'VERSION' => '0.0.8',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/MooseX/Role/AttributeOverride/Meta/Trait/Role/Composite.pm
#
# This file is part of MooseX-Role-AttributeOverride
#
# This software is copyright (c) 2011 by Edward J. Allen III.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use strict; use warnings;
package MooseX::Role::AttributeOverride::Meta::Trait::Role::Composite;
BEGIN {
  $MooseX::Role::AttributeOverride::Meta::Trait::Role::Composite::VERSION = '0.0.8';
}
BEGIN {
  $MooseX::Role::AttributeOverride::Meta::Trait::Role::Composite::AUTHORITY = 'cpan:EALLENIII';
}
# ABSTRACT: Support Role for L<MooseX::Role::AttributeOverride|MooseX::Role::AttributeOverride>
use 5.008;
use utf8;
use Moose::Role;
use Moose::Util::MetaRole;

with 'MooseX::Role::AttributeOverride::Meta::Trait::Role';

around apply_params => sub {
    my $orig = shift;
    my $self = shift;
    $self->$orig(@_);

    ## no critic (ProhibitCallsToUnexportedSubs);

    $self = Moose::Util::MetaRole::apply_metaroles(
        for            => $self,
        role_metaroles => {
            role =>
                ['MooseX::Role::AttributeOverride::Meta::Trait::Role'],
            application_to_class =>
                ['MooseX::Role::AttributeOverride::Meta::Trait::Role::ApplicationToClass'],
            application_to_role =>
                ['MooseX::Role::AttributeOverride::Meta::Trait::Role::ApplicationToRole'],
            },
    );

    ## use critic;

    for my $role (@{$self->get_roles}) {
        if ($role->can('attribute_modifiers')) {
            $self->add_modifiers_from_role($role);
        }
    }

    return $self;
};

no Moose::Role;
1; # Magic true value required at end of module


=pod

=for :stopwords Edward Allen <ealleniii_at_cpan_dot_org> J. III

=encoding utf-8

=head1 NAME

MooseX::Role::AttributeOverride::Meta::Trait::Role::Composite - Support Role for L<MooseX::Role::AttributeOverride|MooseX::Role::AttributeOverride>

=head1 VERSION

  This document describes v0.0.8 of MooseX::Role::AttributeOverride::Meta::Trait::Role::Composite - released June 22, 2011 as part of MooseX-Role-AttributeOverride.

=head1 SYNOPSIS

See L<MooseX::Role::AttributeOverride|MooseX::Role::AttributeOverride>

=head1 DESCRIPTION

Apply this role to composite roles.

=head1 DIAGNOSTICS

See L<MooseX::Role::AttributeOverride|MooseX::Role::AttributeOverride>

=head1 DEPENDENCIES

See L<MooseX::Role::AttributeOverride|MooseX::Role::AttributeOverride>

=head1 INCOMPATIBILITIES

See L<MooseX::Role::AttributeOverride|MooseX::Role::AttributeOverride>

=head1 BUGS AND LIMITATIONS

See L<MooseX::Role::AttributeOverride|MooseX::Role::AttributeOverride>

=head1 SEE ALSO

Please see those modules/websites for more information related to this module.

=over 4

=item *

L<MooseX::Role::AttributeOverride|MooseX::Role::AttributeOverride>

=back

=head1 AUTHOR

Edward Allen <ealleniii_at_cpan_dot_org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Edward J. Allen III.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT
WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER
PARTIES PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND,
EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE
SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME
THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
DAMAGES.

=cut


__END__



lib/MooseX/Role/AttributeOverride.pm
#
# This file is part of MooseX-Role-AttributeOverride
#
# This software is copyright (c) 2011 by Edward J. Allen III.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use strict; use warnings;
package MooseX::Role::AttributeOverride;
BEGIN {
  $MooseX::Role::AttributeOverride::VERSION = '0.0.8';
}
BEGIN {
  $MooseX::Role::AttributeOverride::AUTHORITY = 'cpan:EALLENIII';
}
# ABSTRACT: Allow roles to modify attributes
use 5.008;
use utf8;
use Moose 1.9900 ();
use Moose::Role 1.9900;
use Moose::Exporter;

use MooseX::Role::AttributeOverride::Meta::Trait::Role;
use MooseX::Role::AttributeOverride::Meta::Trait::Role::ApplicationToClass;
use MooseX::Role::AttributeOverride::Meta::Trait::Role::ApplicationToRole;
use MooseX::Role::AttributeOverride::Meta::Trait::Role::Composite;


BEGIN {

    sub has_plus {
        my ($meta, $name, %options) = @_;
        if ($name =~ /\A \+/xms) {
            Moose->throw_error('Do not use a plus prefix with the has_plus sugar');
        }
        if ($meta->can('add_attribute_modifier')) {
            $meta->add_attribute_modifier([ $name => \%options ]);
        }
        else {
            require Moose;
            Moose->throw_error('Attempt to call has_plus on an invalid object');
        }
        return;
    }

    Moose::Exporter->setup_import_methods(
    with_meta => ['has_plus'],
    role_metaroles => {
            role =>
                ['MooseX::Role::AttributeOverride::Meta::Trait::Role'],
            application_to_class =>
                ['MooseX::Role::AttributeOverride::Meta::Trait::Role::ApplicationToClass'],
            application_to_role =>
                ['MooseX::Role::AttributeOverride::Meta::Trait::Role::ApplicationToRole'],
        },
        also => 'Moose::Role',
    );
}



1; # Magic true value required at end of module


=pod

=for :stopwords Edward Allen <ealleniii_at_cpan_dot_org> J. III cpan testmatrix url
annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata
placeholders MooseX MyApp

=encoding utf-8

=head1 NAME

MooseX::Role::AttributeOverride - Allow roles to modify attributes

=head1 VERSION

  This document describes v0.0.8 of MooseX::Role::AttributeOverride - released June 22, 2011 as part of MooseX-Role-AttributeOverride.

=head1 SYNOPSIS

    {
        package MyApp::Role;
        use Moose::Role;
        use MooseX::Role::AttributeOverride;

        has_plus 'fun' => ( default => 'yep', );

        has_plus 'alive' => (
            default => 'yep',
            override_ignore_missing => 1,
        );
    }
    {
        package MyApp::Trait;
        use Moose::Role;
        use MooseX::Role::AttributeOverride;

        has_plus default => (
            default => sub {
                my $attr = shift;
                return sub { $attr->name }
            }
        );
    }
    {
        package MyApp;
        use Moose 1.9900;

        has nolife => (
            is     => 'rw',
            isa    => 'Str',
            traits => ['MyApp::Trait'],
        );

        has 'fun' => (
            is  => 'rw',
            isa => 'Str'
        );

        with qw(MyApp::Role);
    }
    {
        package main;
        use feature 'say';

        my $test = MyApp->new();

        say "I have " . $test->nolife;
        # Says I have nolife
        say "Are you having fun? " . $test->fun;
        # Says Are you having fun? yep

    }

=head1 DESCRIPTION

Moose doesn't allow roles to override attributes using the has '+attr' method.
There are several good reasons for that. Basically, "that's not what a role
is for."  A role is a set of requirements with defaults. A class should
always be able to override a role.

But sometimes you want a role to B<add> features to a class. This is why Moose
has method modifiers. This extension adds attribute modifiers.

=head1 INTERFACE 

=over

=item has_plus

This has exactly the same syntax as the Moose L<has|Moose/has> command, except
you should not use a plus to indicate you are overriding an attribute. 

=item has_plus options

=over

=item override_ignore_missing

Setting this to a true value will allow your role to have modifications to attributes 
that may not exist in the class it is applied to. The default is to die in these
cases.

For example:

    package MyApp::Role;
    use Moose::Role;
    use MooseX::Role::AttributeOverride;

    has_plus 'alive' => (
        default => 'yep',
        override_ignore_missing => 1,
    );

    package MyApp;
    use Moose;

    with qw(MyApp::Role);
    # I'm not dead yet.

The above would not die, even though the MyApp package has no attribute named
'fun.'

=back

=back

=head1 IMPORTANT NOTE

Always apply a role that uses this module B<after> defining attributes.

=head1 META USAGE

This role can be used in traits.  For example, the following works:

    {
        package MyApp::Trait;
        use Moose::Role;
        use MooseX::Role::AttributeOverride;

        has_plus default => (
            default => sub {
                my $attr = shift;
                return sub { $attr->name }
            }
        );
    }
    {
        package MyApp;
        use Moose 1.9900;

        has nolife => (
            is     => 'rw',
            isa    => 'Str',
            traits => ['MyApp::Trait'],
        );

        with qw(MyApp::Role);
    }
    {
        package main;
        use feature 'say';

        my $test = MyApp->new();

        say "I have " . $test->nolife;
        # Says I have nolife
    }

=head1 DIAGNOSTICS

=over

=item Can't find attribute $attr required by $role

You will see this error if your role has an attribute modification for an
attribute that is not in the class. You can squash this by setting the
'override_ignore_missing' option in your 'has_plus' command.

=item Attempt to call has_plus on an invalid object

You really should never see this. Please file a bug report if you do. A test
case would be nice as well.

=item Illegal inherited options 

Moose will throw this error if you try to change an accessor option. See 
L<the Moose manual|Moose::Manual::Attributes/"ATTRIBUTE INHERITANCE"> for more
details.

=item Do not use a plus prefix with the has_plus sugar

There is no need for a plus sign on your attribute:

    # Good
    has 'children', trait => ['good']

    # Bad. Will die.
    has '+children', trait => ['naughty']

=back

=head1 DEPENDENCIES

Moose 1.9900 or newer. Older versions may be supported in a future version of
this module.

=head1 INCOMPATIBILITIES

I am sure that there are some MooseX modules that will not work with this.
Please let me know, and I will at least document them.

=head1 BUGS AND LIMITATIONS

This is not the intended use of roles. As a result, take into account the
following:

=over

=item *

Order matters! If two roles modify the same attribute in the same way,
the second one applied will be the one that is used. This behavior, however,
relies on Moose keeping track of order, which it generally does a good job of,
but no guarantees.

=item *

Currently, the value of the attribute is clobbered when the role is applied.
This may change in the future.

=item *

This works the same as '+has'. This means that you can't override accessor
methods. This is a very sensible Moose limitation.

=item *

After having an issue with Moose, clone_and_inherit_options, and traits that
use _process_options, I reimplemented clone_and_inherit_optiosn in a way that
fixes it.  Sort of.  A side effect of this is that has_plus will not allow you
to override the lazy option, without a default or builder option.

=item *

If you try adding this role before adding the attributes, it won't work.

=back

I am relatively new to Moose. I had an itch, and wrote this Module to scratch
it. Please let me know how to make this module better.

For bugs, test cases are great! 

=head1 SEE ALSO

Please see those modules/websites for more information related to this module.

=over 4

=item *

L<Moose|Moose>

=item *

L<Moose::Manual::Attributes|Moose::Manual::Attributes>

=back

=head1 SUPPORT

=head2 Websites

The following websites have more information about this module, and may be of help to you. As always,
in addition to those websites please use your favorite search engine to discover more resources.

=over 4

=item *

Search CPAN

The default CPAN search engine, useful to view POD in HTML format.

L<http://search.cpan.org/dist/MooseX-Role-AttributeOverride>

=item *

CPAN Testers Dependencies

The CPAN Testers Dependencies is a website that shows a chart of the test results of all dependencies for a distribution.

L<http://deps.cpantesters.org/?module=MooseX::Role::AttributeOverride>

=item *

CPANTS

The CPANTS is a website that analyzes the Kwalitee ( code metrics ) of a distribution.

L<http://cpants.perl.org/dist/overview/MooseX-Role-AttributeOverride>

=item *

CPAN Testers

The CPAN Testers is a network of smokers who run automated tests on uploaded CPAN distributions.

L<http://www.cpantesters.org/distro/M/MooseX-Role-AttributeOverride>

=item *

CPAN Testers Matrix

The CPAN Testers Matrix is a website that provides a visual way to determine what Perls/platforms PASSed for a distribution.

L<http://matrix.cpantesters.org/?dist=MooseX-Role-AttributeOverride>

=back

=head2 Email

You can email the author of this module at C<EALLENIII at cpan.org> asking for help with any problems you have.

=head2 Bugs / Feature Requests

Please report any bugs or feature requests by email to C<bug-moosex-role-attributeoverride at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooseX-Role-AttributeOverride>. You will be automatically notified of any
progress on the request by the system.

=head2 Source Code

The code is open to the world, and available for you to hack on. Please feel free to browse it and play
with it, or whatever. If you want to contribute patches, please send me a diff or prod me to pull
from your repository :)

L<http://github.com/riemann42/MooseX-Role-AttributeOverride>

  git clone git://github.com/riemann42/MooseX-Role-AttributeOverride.git

=head1 AUTHOR

Edward Allen <ealleniii_at_cpan_dot_org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Edward J. Allen III.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT
WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER
PARTIES PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND,
EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE
SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME
THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
DAMAGES.

=cut


__END__

